### PR TITLE
Azure: Use standard sku Public IP addresses

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

As per the bug below, use the standard sku for any public IP address that is created (this allows the use of the standard load balancer etc.)

This branch also updates the azure SKU used to the newest version.

## QA steps

When you deploy any machines to an Azure provider, using the Azure we UI inspect the created public IP addresses in the resources viewer and see the SKU states 'standard' (as opposed to 'basic').

## Documentation changes
 n/a

## Bug reference

https://bugs.launchpad.net/juju/+bug/1799238

### CODE REVIEW

File: provider/azure/environ_test.go [hunk 1]
```diff
@@ -980,8 +1012,10 @@ func (s *environSuite) assertStartInstanceRequests(
 		Location:   "westus",
 		Tags:       to.StringMap(s.vmTags),
 		Properties: &network.PublicIPAddressPropertiesFormat{
-			PublicIPAllocationMethod: network.Dynamic,
+			PublicIPAllocationMethod: network.Static,
```
[the remark this answers is not in the document]
Comment (Veebers) [reply provider/azure/environ_test.go]: yep, that reflects the crux of this pr which is to move to standard sku which only supports static IPs.

File: provider/azure/internal/azureresources/resources.go [hunk 2]
```diff
@@ -20,84 +25,80 @@ import (
...
 }
 
 // GetByID gets a resource by ID.
 //
-// See: resources.GroupClient.GetByID.
-func (client ResourcesClient) GetByID(resourceID, apiVersion string) (result resources.GenericResource, err error) {
-	req, err := client.GetByIDPreparer(resourceID)
+// See: resources.Client.GetByID.
+func (client ResourcesClient) GetByID(ctx context.Context, resourceID, apiVersion string) (result resources.GenericResource, err error) {
+	if tracing.IsEnabled() {
```
[the remark this answers is not in the document]
Comment (Veebers) [reply provider/azure/internal/azureresources/resources.go]: The tracing? No its something the sdk does. These functions are straight copies of code from the sdk with the addition of the setAPIVersion call.

File: provider/azure/environ.go [hunk 3]
```diff
@@ -1267,9 +1302,17 @@ func (env *azureEnviron) deleteVirtualMachine(
 	for _, nic := range networkInterfaces {
 		nicName := to.String(nic.Name)
 		logger.Tracef("deleting NIC %q", nicName)
-		resultCh, errCh := nicClient.Delete(env.resourceGroup, nicName, nil)
-		if result, err := <-resultCh, <-errCh; err != nil {
-			if !isNotFoundResponse(result) {
+		nicFuture, err := nicClient.Delete(ctx, env.resourceGroup, nicName)
```
[the remark this answers is not in the document]
Comment (Veebers) [reply provider/azure/environ.go]: As discussed in IRC for any future reference. The different Result() methods take concrete struct types which means we can't make this any less boiler-platy

File: provider/azure/environ.go [hunk 4]
```diff
@@ -257,10 +262,12 @@ func (env *azureEnviron) initResourceGroup(controllerUUID string, controller boo
...
 	env.mu.Unlock()
 
+	sanitisedTags := to.StringMapPtr(tags)
 	logger.Debugf("creating resource group %q", env.resourceGroup)
-	if _, err := resourceGroupsClient.CreateOrUpdate(env.resourceGroup, resources.Group{
+	sdkCtx := stdcontext.Background()
+	if _, err := resourceGroupsClient.CreateOrUpdate(sdkCtx, env.resourceGroup, resources.Group{
 		Location: to.StringPtr(env.location),
-		Tags:     to.StringMapPtr(tags),
+		Tags:     *sanitisedTags,
```
Comment (wallyworld) [inline provider/azure/environ.go]: We can just leave the original code as is right?

File: provider/azure/environ.go [hunk 5]
```diff
@@ -1110,11 +1123,12 @@ func (env *azureEnviron) StopInstances(ctx context.ProviderCallContext, ids ...i
 	cancelResults := make([]error, len(ids))
 	for i, id := range ids {
 		logger.Debugf("canceling deployment for instance %q", id)
+		sdkCtx := stdcontext.Background()
```
Comment (wallyworld) [inline provider/azure/environ.go]: This needs to be done inside the go func below, just after the defer wg.Done() line.

File: provider/azure/environ.go [hunk 6]
```diff
@@ -1161,10 +1175,12 @@ func (env *azureEnviron) StopInstances(ctx context.ProviderCallContext, ids ...i
 		}
 		// The deployment does not exist, so there's nothing more to do.
 		logger.Debugf("deleting instance %q", id)
+		sdkCtx := stdcontext.Background()
```
Comment (wallyworld) [inline provider/azure/environ.go]: ditto etc

File: provider/azure/internal/azureresources/resources.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply provider/azure/internal/azureresources/resources.go]: Why do we need to copy SDK code? Isn't the point of this new SDK that the version to use is now determined by the import path of the various bit used? So we don't need to send anything down the wire ourselves.

File: provider/azure/internal/azureresources/resources.go [hunk 7]
```diff
@@ -4,13 +4,18 @@
...
-	"github.com/Azure/azure-sdk-for-go/arm/resources/resources"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
 	"github.com/Azure/go-autorest/autorest"
 	"github.com/Azure/go-autorest/autorest/validation"
+	"github.com/Azure/go-autorest/tracing"
 )
 
+// The package's fully qualified name.
+const fqdn = "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+
```
Comment (wallyworld) [inline provider/azure/internal/azureresources/resources.go]: typo in comment below s/GroupClient/Client

File: provider/azure/storage.go [hunk 8]
```diff
@@ -907,7 +949,7 @@ func getStorageAccountKey(
 	resourceGroup, accountName string,
 ) (*armstorage.AccountKey, error) {
 	logger.Debugf("getting keys for storage account %q", accountName)
-	listKeysResult, err := client.ListKeys(resourceGroup, accountName)
+	listKeysResult, err := client.ListKeys(context.Background(), resourceGroup, accountName)
```
Comment (wallyworld) [inline provider/azure/storage.go]: stdCtx?

File: provider/azure/credentials.go [hunk 9]
```diff
@@ -223,7 +225,7 @@ func (c environProviderCredentials) azureCLICredential(
 	}
 	params.ResourceManagerAuthorizer = autorest.NewBearerAuthorizer(resourceManagerAuthorizer.Token())
 
-	applicationId, password, err := c.servicePrincipalCreator.Create(params)
+	applicationId, password, err := c.servicePrincipalCreator.Create(context.Background(), params)
```
Comment (wallyworld) [inline provider/azure/credentials.go]: stdCtx?

File: provider/azure/credentials.go [hunk 10]
```diff
@@ -248,7 +250,7 @@ func (c environProviderCredentials) accountCredential(
 	if err != nil {
 		return cloud.Credential{}, errors.Annotatef(err, "cannot get access token for %s", acc.ID)
 	}
-	applicationId, password, err := c.servicePrincipalCreator.Create(azureauth.ServicePrincipalParams{
+	applicationId, password, err := c.servicePrincipalCreator.Create(context.Background(), azureauth.ServicePrincipalParams{
```
Comment (wallyworld) [inline provider/azure/credentials.go]: ditto

File: provider/azure/networking.go [hunk 11]
```diff
@@ -246,7 +247,7 @@ func networkSecurityRules(
 	nsgClient network.SecurityGroupsClient,
 	resourceGroup string,
 ) ([]network.SecurityRule, error) {
-	nsg, err := nsgClient.Get(resourceGroup, internalSecurityGroupName, "")
+	nsg, err := nsgClient.Get(context.Background(), resourceGroup, internalSecurityGroupName, "")
```
Comment (wallyworld) [inline provider/azure/networking.go]: ditto

File: provider/azure/environ_test.go [hunk 12]
```diff
@@ -1582,7 +1582,7 @@ func (s *environSuite) TestDestroyControllerErrors(c *gc.C) {
 	s.sender = azuretesting.Senders{
 		s.makeSender(".*/resourcegroups", result), // GET
 		makeErrorSender("foo"),                    // DELETE
-		makeErrorSender("bar"),                    // DELETE
+		makeErrorSender("foo"),                    // DELETE
```
Comment (wallyworld) [inline provider/azure/environ_test.go]: the idea was to have different errors, so needs to remain as "bar"

**Comment by Veebers** [comment]
$$merge$$

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**